rocketpy/mathutils/function.py

```diff
@@ -333,141 +338,163 @@ def set_extrapolation(self, method="constant"):
-                        x_interval = 1 if x < x_min else -1
-                        x_left = x_data[x_interval - 1]
-                        y_left = y_data[x_interval - 1]
-                        dx = float(x_data[x_interval] - x_left)
-                        dy = float(y_data[x_interval] - y_left)
-                        y = (x - x_left) * (dy / dx) + y_left
-                    else:  # Extrapolation is set to constant
-                        y = y_data[0] if x < x_min else y_data[-1]
-                return y
-
-        elif self.__interpolation__ == "akima":
-            coeffs = np.array(self.__akima_coefficients__)
-
-            def get_value_opt(x):
-                if x_min <= x <= x_max:
-                    # Interpolate
-                    x_interval = np.searchsorted(x_data, x)
-                    x_interval = x_interval if x_interval != 0 else 1
-                    a = coeffs[4 * x_interval - 4 : 4 * x_interval]
-                    y = a[3] * x**3 + a[2] * x**2 + a[1] * x + a[0]
-                else:
-                    # Extrapolate
-                    if extrapolation == 0:  # Extrapolation == zero
-                        y = 0
-                    elif extrapolation == 1:  # Extrapolation == natural
-                        a = coeffs[:4] if x < x_min else coeffs[-4:]
-                        y = a[3] * x**3 + a[2] * x**2 + a[1] * x + a[0]
-                    else:  # Extrapolation is set to constant
-                        y = y_data[0] if x < x_min else y_data[-1]
-                return y
-
-        elif self.__interpolation__ == "polynomial":
-            coeffs = self.__polynomial_coefficients__
-
-            def get_value_opt(x):
-                # Interpolate... or extrapolate
-                if x_min <= x <= x_max:
-                    # Interpolate
-                    y = np.sum(coeffs * x ** np.arange(len(coeffs)))
-                else:
-                    # Extrapolate
-                    if extrapolation == 0:  # Extrapolation == zero
-                        y = 0
-                    elif extrapolation == 1:  # Extrapolation == natural
-                        y = 0
-                        for i, coef in enumerate(coeffs):
-                            y += coef * (x**i)
-                    else:  # Extrapolation is set to constant
-                        y = y_data[0] if x < x_min else y_data[-1]
-                return y
-
-        elif self.__interpolation__ == "shepard":
-            # change the function's name to avoid mypy's error
-            def get_value_opt_multiple(*args):
-                return self.__interpolate_shepard__(args)
-
-            get_value_opt = get_value_opt_multiple
+        coeffs = self._coeffs
+        if x_min <= x <= x_max:
+            y = self._interpolation_func(x, x_min, x_max, x_data, y_data, coeffs)
+        else:
+            y = self._extrapolation_func(x, x_min, x_max, x_data, y_data, coeffs)
+        return y
```

Comment:
Excellent. When thinking about `Function` class refactors, one of the goals was making this method look like this.